File: 2707-extra-characters-in-a-string/2707-extra-characters-in-a-string.ts
function minExtraChar(s: string, dictionary: string[]): number {
  const n = s.length,
        dict = new Set(dictionary),
        mem = new Map<number, number>();

  const dp = (l: number): number => {
    if (l === n) return 0;
    if (mem.has(l)) return mem.get(l);

    let ans = dp(l + 1) + 1;
    for (let r = l; r < n; r++) {
      if (dict.has(s.slice(l, r + 1))) ans = Math.min(ans, dp(r + 1));
    }

    mem.set(l, ans);
    return ans
  };

  return dp(0);
};

File: 2707-extra-characters-in-a-string/2707-extra-characters-in-a-string.py
class Solution:
  def minExtraChar(self, s: str, words: List[str]) -> int:
    pos = []
    for word in words:
      i = 0
      while True:
        i = s.find(word, i)
        if i == -1: break
        pos.append((i, i + len(word)))
        i += 1
    pos.sort()

    mem = {}
    def traverse(i: int, j: int, val: int) -> int:
      mem[j] = val
      for i in range(i, len(pos)):
        l, r = pos[i]
        if j <= l and (r not in mem or mem[r] > mem[j] - (r - l)):
          val = min(val, traverse(i + 1, r, mem[j] - (r - l)))
      return val

    return traverse(0, 0, len(s))